slackclient/client.py

```diff
@@ -151,19 +173,39 @@ def api_call(self, method, timeout=None, **kwargs):
 
             See here for more information on responses: https://api.slack.com/web
         """
-        response_body = self.server.api_call(method, timeout=timeout, **kwargs)
+        # Check for missing or expired access token before submitting the request
+        if method != 'oauth.access' and self.refresh_token:
+            current_ts = int(time.time()) * 1000
+            expired_token = current_ts > self.access_token_expires_at
+            if (expired_token or self.token is None):
+                self.refresh_access_token()
+
+        response_body = self.server.api_call(self.token, request=method, timeout=timeout, **kwargs)
+
+        # Attempt to parse the response as JSON
         try:
             result = json.loads(response_body)
         except ValueError as json_decode_error:
             raise ParseResponseError(response_body, json_decode_error)
+        response_json = json.loads(response_body)
 
-        if "ok" in result and result["ok"]:
+        if result["ok"]:
             if method == 'im.open':
                 self.server.attach_channel(kwargs["user"], result["channel"]["id"])
             elif method in ('mpim.open', 'groups.create', 'groups.createchild'):
                 self.server.parse_channel_data([result['group']])
             elif method in ('channels.create', 'channels.join'):
                 self.server.parse_channel_data([result['channel']])
+        else:
+            # if the API request returns an invalid_auth error, refresh the token and try again
+            if 'retry' in kwargs:
+                raise SlackClientError("Request failing after token refresh")
+
+            elif (self.refresh_token and response_json['error'] == 'invalid_auth'):
+                    # If the API returns 'ok' false, attempt to refresh the client's access token
+                    self.refresh_access_token()
+                    # If token refresh was successful, retry the original API request
+                    return self.api_call(method, timeout, retry=True, **kwargs)
```

Comment:
Correct me if I'm wrong, but I believe `retry` is necessary because this method is recursive. It requires this key to raise an exception on the subsequent request.

Reply:
it is recursive, but it only enters this condition if the error was an `invalid_auth` error. and if that `invalid_auth` error couldn't be corrected by refreshing the token, then the statement before this one (`self.refresh_access_token()`) would raise an error, and we would never recurse. i don't see any situation where we might recurse more than once.

Reply:
Yep, you're right! I see that now. I've removed it.